### Importing Library:

In [1]:
import pandas as pd

### Loading "Matches" Dataset:

In [2]:
data1 = pd.read_csv("matches.csv")
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
data2 = pd.read_csv('deliveries.csv')
data1.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

### Filtering RR and CSK matches only:

In [3]:
csk_rr = data1.loc[((data1['team1'] == 'Rajasthan Royals') & (data1['team2'] == 'Chennai Super Kings') | (data1['team1'] == 'Chennai Super Kings') & (data1['team2'] == 'Rajasthan Royals'))]
csk_rr.head(10)

id  season       city        date                team1  \
82    83    2008     Jaipur  2008-05-04  Chennai Super Kings   
109  110    2008    Chennai  2008-05-24     Rajasthan Royals   
116  117    2008     Mumbai  2008-06-01  Chennai Super Kings   
136  137    2009  Centurion  2009-04-30  Chennai Super Kings   
151  152    2009  Kimberley  2009-05-09     Rajasthan Royals   
197  198    2010  Ahmedabad  2010-03-28     Rajasthan Royals   
205  206    2010    Chennai  2010-04-03  Chennai Super Kings   
275  276    2011    Chennai  2011-05-04     Rajasthan Royals   
284  285    2011     Jaipur  2011-05-09  Chennai Super Kings   
333  334    2012    Chennai  2012-04-21     Rajasthan Royals   

                   team2          toss_winner toss_decision  result  \
82      Rajasthan Royals  Chennai Super Kings           bat  normal   
109  Chennai Super Kings     Rajasthan Royals           bat  normal   
116     Rajasthan Royals     Rajasthan Royals         field  normal   
136     Rajasthan Royals     Rajasthan Royals         field  normal   
151  Chennai Super Kings     Rajasthan Royals           bat  normal   
197  Chennai Super Kings     Rajasthan Royals           bat  normal   
205     Rajasthan Royals  Chennai Super Kings           bat  normal   
275  Chennai Super Kings     Rajasthan Royals           bat  normal   
284     Rajasthan Royals     Rajasthan Royals         field  normal   
333  Chennai Super Kings     Rajasthan Royals           bat  normal   

     dl_applied               winner  win_by_runs  win_by_wickets  \
82            0     Rajasthan Royals            0               8   
109           0     Rajasthan Royals           10               0   
116           0     Rajasthan Royals            0               3   
136           0  Chennai Super Kings           38               0   
151           0  Chennai Super Kings            0               7   
197           0     Rajasthan Royals           17               0   
205           0  Chennai Super Kings           23               0   
275           0  Chennai Super Kings            0               8   
284           0  Chennai Super Kings           63               0   
333           0  Chennai Super Kings            0               7   

    player_of_match                            venue      umpire1  \
82    Sohail Tanvir           Sawai Mansingh Stadium    Asad Rauf   
109       JA Morkel  MA Chidambaram Stadium, Chepauk    DJ Harper   
116       YK Pathan       Dr DY Patil Sports Academy    BF Bowden   
136        SK Raina                  SuperSport Park   GAV Baxter   
151     S Badrinath            De Beers Diamond Oval   GAV Baxter   
197         NV Ojha     Sardar Patel Stadium, Motera    SS Hazare   
205         M Vijay  MA Chidambaram Stadium, Chepauk  RE Koertzen   
275      MEK Hussey  MA Chidambaram Stadium, Chepauk    SS Hazare   
284         M Vijay           Sawai Mansingh Stadium  K Hariharan   
333    F du Plessis  MA Chidambaram Stadium, Chepauk    Aleem Dar   

             umpire2 umpire3  
82    AV Jayaprakash     NaN  
109       SL Shastri     NaN  
116      RE Koertzen     NaN  
136      RE Koertzen     NaN  
151  HDPK Dharmasena     NaN  
197       SJA Taufel     NaN  
205        RB Tiffin     NaN  
275        RB Tiffin     NaN  
284       SJA Taufel     NaN  
333     BNJ Oxenford     NaN

## Question 1: Who will win the match?

#### Solution: Checking counts of winner:

In [4]:
csk_rr['winner'].value_counts()/csk_rr.shape[0] * 100

Chennai Super Kings    66.666667
Rajasthan Royals       33.333333
Name: winner, dtype: float64

**Answer:** As per the above counts, CSK won more matches (**66.67 %**) against SRH. I would predict that CSK will win today's match.

## Question 2: How many runs will RR score in first 6 overs?

### Loading "Deliveries" dataset:

In [5]:
data2 = pd.read_csv("deliveries.csv")
data2.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  wide_runs  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0          0         0   
1     2  DA Warner    S Dhawan  TS Mills              0          0         0   
2     3  DA Warner    S Dhawan  TS Mills              0          0         0   
3     4  DA Warner    S Dhawan  TS Mills              0          0         0   
4     5  DA Warner    S Dhawan  TS Mills              0          2         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs player_dismissed dismissal_kind fielder  
0           0              NaN            NaN     NaN  
1           0              NaN            NaN     NaN  
2           4              NaN            NaN     NaN  
3           0              NaN            NaN     NaN  
4           2              NaN            NaN     NaN

#### Solution: Filtering data with RR as batting team and CSK as bowling team. Also getting data beween 1 and 6 overs only:

In [6]:
rr_runs = data2.loc[(data2['batting_team'] == "Rajasthan Royals") & (data2['bowling_team'] == "Chennai Super Kings")]
rr_runs = rr_runs.loc[(rr_runs['over'] >= 1) & (rr_runs['over'] <= 6)]
rr_runs = pd.DataFrame(rr_runs.groupby('match_id')['total_runs'].sum())
rr_runs = rr_runs.rename({'total_runs' : 'Powerplay Runs'}, axis = 1)
rr_runs

Powerplay Runs
match_id                
83                    44
110                   67
117                   41
137                   38
152                   34
198                   58
206                   62
276                   47
285                   48
334                   41
361                   28
410                   55
442                   24
467                   38
494                   42
533                   47
563                   40
7910                  37
7936                  58
11148                 48
11316                 60

#### Getting mean of the powerplay runs match wise:

In [7]:
avg_pp_runs = rr_runs['Powerplay Runs'].mean()
avg_pp_runs

45.57142857142857

**Answer:** Average runs in each match in first 6 overs is **46** for RR.

## Question 3: What will Deepak Chahar's economy rate be?

#### Solution: Observing match wise Deepak Chahr's economy rate by dividing runs by over:

In [8]:
# Filtering data for Deepk Chahar as bowler:
deepak_runs = data2.loc[data2['bowler'] == "DL Chahar"]
economy_rate = pd.DataFrame(deepak_runs.groupby('match_id')['total_runs'].sum())/4
economy_rate = economy_rate.rename({'total_runs' : 'economy_rate'}, axis = 1)
economy_rate

economy_rate
match_id              
2                 5.25
9                 9.00
33                4.50
625               3.25
629               7.25
7894              3.50
7898              4.75
7905              9.50
7910              8.50
7913              3.75
7917              5.00
7920              5.00
7939              5.75
7945              6.25
7949              8.75
7950              8.25
7953              6.75
11137             4.50
11141             5.25
11148             5.00
11151             5.75
11309            10.75
11314             5.75
11316             8.75
11320             9.00
11324             7.75
11330             7.00
11332             8.00
11335            11.75
11341             8.00
11346             7.00
11412             7.50
11414             7.25
11415             6.75

#### Finding average economy rate in each match:

In [9]:
avg_er = economy_rate['economy_rate'].mean()
avg_er

6.786764705882353

**Answer:** As per the above analysis, average per match economy rate of Deepak Chahr is **6.79**.

## Question 4: How many wickets will Jofra Archer take in the death?

#### Solution: Filterng data where bowler is Jofra Archer and overs are 16-20 (Death Overs):

In [10]:
ja_wickets = data2.loc[(data2['bowler'] == "J Archer") & ((data2['over'] >= 16) & (data2['over'] <= 20))]
ja_wickets = pd.DataFrame(ja_wickets.groupby('match_id')['player_dismissed'].count())
ja_wickets = ja_wickets.rename({'player_dismissed' : 'No. of Wickets'}, axis = 1)
ja_wickets

No. of Wickets
match_id                
7914                   3
7921                   3
7925                   1
7931                   0
7933                   0
7936                   0
7940                   0
7942                   0
7946                   0
7951                   0
11140                  0
11144                  0
11148                  1
11150                  1
11316                  0
11318                  2
11323                  2
11327                  1
11331                  0
11334                  0

#### Observing average wickets taken by Jofra Archer in death overs:

In [11]:
avg_wickets = ja_wickets['No. of Wickets'].mean()
avg_wickets

0.7

**Answer:** As per my analysis, average wickets taken by Jofra Archer in death overs are **0.7**, which means **0-1 wickets** .

## Question 5: What will Steve Smith's strike rate be during the match?

#### Solution: Filtering data with Steve Smith as batsman and calculating average per match strike rate:

In [12]:
ss = data2.loc[data2['batsman'] == 'SPD Smith']
ss_balls = pd.DataFrame(ss.groupby('match_id')['ball'].count())
ss_runs = pd.DataFrame(ss.groupby('match_id')['batsman_runs'].sum())
ss_data = pd.merge(ss_balls, ss_runs, on='match_id')
ss_data['Strike Rate'] = 100 * ss_data['batsman_runs']/ss_data['ball']
ss_data = ss_data.rename({'ball' : 'Balls', 'batsman_runs' : 'Runs'}, axis = 1)
ss_data

Balls  Runs  Strike Rate
match_id                          
2            55    84   152.727273
4            28    26    92.857143
13           30    43   143.333333
17           26    27   103.846154
25           21    27   128.571429
28           12    17   141.666667
29           37    51   137.837838
33           32    45   140.625000
38            2     4   200.000000
40            9     9   100.000000
43           40    34    85.000000
51           35    38   108.571429
54           18    15    83.333333
56            2     1    50.000000
59           53    51    96.226415
310          32    39   121.875000
315          12    25   208.333333
321          13    13   100.000000
324          23    44   191.304348
329          14    16   114.285714
332          22    23   104.545455
335          14    34   242.857143
341          14    26   185.714286
348          27    47   174.074074
351           6     2    33.333333
353          16    14    87.500000
358          23    18    78.260870
362          25    24    96.000000
365          31    37   119.354839
399          16    39   243.750000
402          25    17    68.000000
409           9     6    66.666667
411          32    41   128.125000
419           8    17   212.500000
422          40    35    87.500000
426           4     4   100.000000
464          15    27   180.000000
467          20    19    95.000000
476          14    24   171.428571
482           2     3   150.000000
487          34    22    64.705882
492          22    48   218.181818
494           8     9   112.500000
520          24    33   137.500000
523           9    10   111.111111
526          54    79   146.296296
529          11    13   118.181818
533           7     6    85.714286
536           3     5   166.666667
540          28    31   110.714286
549          22    23   104.545455
558          40    68   170.000000
563           6     4    66.666667
570          11    14   127.272727
574          13    12    92.307692
582           6     5    83.333333
586          27    38   140.740741
592           4     4   100.000000
596          28    31   110.714286
598          41    46   112.195122
601          54   101   187.037037
605          24    45   187.500000
11140        17    24   141.176471
11148        33    31    93.939394
11150        32    39   121.875000
11312        61    79   129.508197
11316        22    15    68.181818
11318        15    12    80.000000
11327        48    59   122.916667
11331        33    51   154.545455
11334         6     2    33.333333
11336        16    22   137.500000

### Average Strike rate per match:

In [13]:
avg_sr = ss_data['Strike Rate'].mean()
avg_sr

124.47036392976513

**Answer:** As per the above analysis, we can see that on an average Steve Smith keeps strike rate of **124 runs** per match.